pyteal/ast/router.py

```diff
@@ -0,0 +1,610 @@
+    )
+
+    @staticmethod
+    def never() -> "OnCompleteAction":
+        return OnCompleteAction()
+
+    @staticmethod
+    def create_only(
+        f: Expr | SubroutineFnWrapper | ABIReturnSubroutine,
+    ) -> "OnCompleteAction":
+        return OnCompleteAction(on_create=f)
+
+    @staticmethod
+    def call_only(
+        f: Expr | SubroutineFnWrapper | ABIReturnSubroutine,
+    ) -> "OnCompleteAction":
+        return OnCompleteAction(on_call=f)
+
+    @staticmethod
+    def always(
+        f: Expr | SubroutineFnWrapper | ABIReturnSubroutine,
+    ) -> "OnCompleteAction":
+        return OnCompleteAction(on_create=f, on_call=f)
+
+    def is_empty(self) -> bool:
+        return not (self.on_call or self.on_create)
+
+
+OnCompleteAction.__module__ = "pyteal"
+
+
+@dataclass(frozen=True)
+class BareCallActions:
+    """
+    OnCompletion Actions keep track of bare-call registrations to all OnCompletion cases.
+    """
+
+    close_out: OnCompleteAction = field(kw_only=True, default=OnCompleteAction.never())
+    clear_state: OnCompleteAction = field(
+        kw_only=True, default=OnCompleteAction.never()
+    )
+    delete_application: OnCompleteAction = field(
+        kw_only=True, default=OnCompleteAction.never()
+    )
+    no_op: OnCompleteAction = field(kw_only=True, default=OnCompleteAction.never())
+    opt_in: OnCompleteAction = field(kw_only=True, default=OnCompleteAction.never())
+    update_application: OnCompleteAction = field(
+        kw_only=True, default=OnCompleteAction.never()
+    )
+
+    def is_empty(self) -> bool:
+        for action_field in fields(self):
+            action: OnCompleteAction = getattr(self, action_field.name)
+            if not action.is_empty():
+                return False
+        return True
+
+    def approval_construction(self) -> Optional[Expr]:
+        oc_action_pair: dict[EnumInt, OnCompleteAction] = {
+            OnComplete.NoOp: self.no_op,
+            OnComplete.OptIn: self.opt_in,
+            OnComplete.CloseOut: self.close_out,
+            OnComplete.UpdateApplication: self.update_application,
+            OnComplete.DeleteApplication: self.delete_application,
+        }
+        if all(oca.is_empty() for oca in oc_action_pair.values()):
+            return None
+        conditions_n_branches: list[CondNode] = list()
+        for oc, oca in oc_action_pair.items():
+            if oca.on_call:
+                conditions_n_branches.append(
+                    CondNode(
+                        And(Txn.on_completion() == oc, Txn.application_id() != Int(0)),
+                        ASTBuilder.wrap_handler(False, oca.on_call),
+                    )
+                )
+            if oca.on_create:
+                conditions_n_branches.append(
+                    CondNode(
+                        And(Txn.on_completion() == oc, Txn.application_id() == Int(0)),
+                        ASTBuilder.wrap_handler(False, oca.on_create),
+                    )
+                )
```

Comment:
Assuming you adopt my suggestion from https://github.com/algorand/pyteal/pull/170#discussion_r882123067, this code should be changed so that it doesn't duplicate the `oca` handlers.

Something like this:
```suggestion
            # assume properties oca.handler and oca.config
            wrapped_handler = ASTBuilder.wrap_handler(False, oca.handler)
            match oca.config:
                case CallConfig.NEVER:
                    continue
                case CallConfig.ALL:
                    cond_body = wrapped_handler
                case CallConfig.CALL:
                    cond_body = Seq(Assert(Txn.application_id() != Int(0)), wrapped_handler)
                case CallConfig.CREATE:
                    cond_body = Seq(Assert(Txn.application_id() == Int(0)), wrapped_handler)
                case _:
                    raise TealInternalError(f"Unexpected value for CallConfig: {oca.config}")
            
            conditions_n_branches.append(
                CondNode(
                    Txn.on_completion() == oc,
                    cond_body,
                )
            )
```

Reply:
correct, if I walk down that path, then there is no point in discussion by cases like previous code snippet.